README.rst
About conn-check
================

conn-check allows for checking connectivity with external services.

You can write a config file that defines services that you need to
have access to, and conn-check will check connectivity with each.

It supports various types of services, all of which allow for
basic network checks, but some allow for confirming credentials
work also.

Configuration
-------------

The configuration is done via a yaml file. The file defines a list
of checks to do:

.. code-block:: yaml

    - type: tcp
      host: localhost
      port: 80
    - type: tls
      host: localhost
      port: 443
      disable_tls_verification: false

Each check defines a type, and then options as appropriate for that type.

For a step by step guide on configuring conn-check for your application
`see the tutorial <http://conn-check.readthedocs.org/>`_.

Check Types
-----------

tcp
```

A simple tcp connectivity check.

host
    The host.

port
    The port.

timeout
    Optional connection timeout in seconds. Default: 10 (or value from ``--connect-timeout``).


tls
```

A check that uses TLS (`ssl` is a deprecated alias for this type).

host
    The host.

port
    The port.

disable_tls_verification
    Optional flag to disable verification of TLS certs and handshake. Default:
    false.

timeout
    Optional connection timeout in seconds. Default: 10 (or value from ``--connect-timeout``).


udp
```

Check that sending a specific UDP packet gets a specific response.

host
    The host.

port
    The port.

send
    The string to send.

expect
    The string to expect in the response.

timeout
    Optional connection timeout in seconds. Default: 10 (or value from ``--connect-timeout``).


http
````

Check that a HTTP/HTTPS request succeeds (`https` also works).

url
    The URL to fetch.

method
    Optional HTTP method to use. Default: "GET".

expected_code
    Optional status code that defines success. Default: 200.

proxy_url
    Optional HTTP/HTTPS proxy URL to connect via, including protocol,
    if set proxy_{host,port} are ignored.

proxy_host
    Optional HTTP/HTTPS proxy to connect via.

proxy_port
    Optional port to use with ``proxy_host``. Default: 8000.

headers:
    Optional headers to send, as a dict of key-values. Multiple values can be
    given as a list/tuple of lists/tuples, e.g.:
    ``[('foo', 'bar'), ('foo', 'baz')]``

body:
    Optional raw request body string to send.

disable_tls_verification:
    Optional flag to disable verification of TLS certs and handshake. Default:
    false.

timeout
    Optional connection timeout in seconds. Default: 10 (or value from ``--connect-timeout``).

allow_redirects
    Optional flag to Follow 30x redirects. Default: false.

params
    Optional dict of params to URL encode and pass in the querystring.

cookies
    Optional dict of cookies to pass in the request headers.

auth
    Optional `basic HTTP auth <https://en.wikipedia.org/wiki/Basic_access_authentication>`_
    credentials, as a tuple/list: ``(username, password)``.

digest_auth
    Optional `digest HTTP auth <https://en.wikipedia.org/wiki/Digest_access_authentication>`_
    credentials, as a tuple/list: ``(username, password)``.


amqp
````

Check that an AMQP server can be authenticated against.

host
    The host.

port
    The port.

username
    The username to authenticate with.

password
    The password to authenticate with.

use_tls
    Optional flag whether to connect with TLS. Default: true.

vhost
    Optional vhost name to connect to. Default '/'.

timeout
    Optional connection timeout in seconds. Default: 10 (or value from ``--connect-timeout``).


postgres
````````

Check that a PostgreSQL db can be authenticated against (`postgresql` also works).

host
    The host.

port
    The port.

username
    The username to authenticate with.

password
    The password to authenticate with.

database
    The database to connect to.

timeout
    Optional connection timeout in seconds. Default: 10 (or value from ``--connect-timeout``).


redis
`````

Check that a redis server is present, optionally checking authentication.

host
    The host.

port
    The port.

password
    Optional password to authenticatie with.

timeout
    Optional connection timeout in seconds. Default: 10 (or value from ``--connect-timeout``).


memcache
````````

Check that a memcached server is present (`memcached` also works).

host
    The host.

port
    The port.

timeout
    Optional connection timeout in seconds. Default: 10 (or value from ``--connect-timeout``).


mongodb
```````

Check that a MongoDB server is present (`mongo` also works).

host
    The host.

port
    Optional port. Default: 27017.

username
    Optional username to authenticate with.

password
    Optional password to authenticate with.

database
    Optional database name to connect to, if not set the ``test`` database will be used,
    if this database does not exist (or is not available to the user) you will need to
    provide a database name.

timeout
    Optional connection timeout in seconds. Default: 10 (or value from ``--connect-timeout``).


smtp
````

Check that we can reach, authenticate with and send an email using an SMTP server.

**Note 1**: if this check succeeds an email is actually sent to the email
defined in ``to_address``, be careful how this is check is configured so it doesn't
unintentionally spam anyone.

**Note 2**: only EHLO/HELO over a TLS connection is supported with the ``use_tls``
flag, this check cannot currently create new TLS connection using the
`STARTTLS Extension <https://tools.ietf.org/html/rfc3207>`_.

host
    The host.

port
    The port, normally 465 for TLS and 25 for plaintext.

username
    Username to authenticate with.

password
    Password to authenticate with.

from_address:
    Email address to send `from`.

to_address:
    Email address to send `to`.

message:
    Optional email body.

subject:
    Optional email subject.

helo_fallback:
    Optional flag that defines whether to fall back to ``HELO`` if the ``EHLO``
    extended command set fails.

use_tls:
    Optional flag to enable TLS security on connection. Default: true.

timeout
    Optional connection timeout in seconds. Default: 10 (or value from ``--connect-timeout``).


Tags
----

Every check type also supports a ``tags`` field, which is a list of tags that
can be used with the ``--include-tags`` and ``--exclude-tags`` arguments to conn-check.

Example YAML:

.. code-block:: yaml

    - type: http
      url: http://google.com/
      tags:
        - external

To run just "external" checks::

    conn-check --include-tags=external ...

To run all the checks *except* external::

    conn-check --exclude-tags=external

Buffered/Ordered output
-----------------------

conn-check normally executes with output to ``STDOUT`` buffered so that the output can be ordered,
with failed checks being printed first, grouping by destination etc.

If you'd rather see results as they available you can use the ``-U``/``--unbuffered-output`` option
to disable buffering.

Generating firewall rules
-------------------------

conn-check includes the ``conn-check-export-fw`` utility which takes the same arguments as
``conn-check`` but runs using ``--dry-run`` mode and outputs a set of `egress` firewall
rules in an easy to parse YAML format, for example:

.. code-block:: yaml

    # Generated from the conn-check demo.yaml file
    egress:
    - from_host: mydevmachine
      ports: [8080]
      protocol: udp
      to_host: localhost
    - from_host: mydevmachine
      ports: [80, 443]
      protocol: tcp
      to_host: login.ubuntu.com
    - from_host: mydevmachine
      ports: [6379, 11211]
      protocol: tcp
      to_host: 127.0.0.1

You can then use this output to generate your environments firewall rules (e.g. with
`EC2 security groups`, `OpenStack Neutron`, `iptables` etc.).

``conn-check-convert-fw`` is a utility that does just this, it accepts multiple firewall
rule YAML files, merges/de-dupes them, and outputs commands for AWS, Openstack Neutron,
OpenStack Nova (client), iptables, and ufw (mostly for testing purposes).

It is designed for this workflow:

 * On each host you run conn-check from, you run ``conn-check-export-fw`` to generate
   a YAML file containing egress firewall rules.
 * Each of these files is transfered to a host with the correct DNS entries for the
   egress hosts.
 * On this host ``conn-check-convert-fw`` is run to generate a set of commands
   for your firewall.
 * These commands are audited by a human / possibly merged with other rules, such as
   adding ingress rules, and then run to update your environment's firewall.

Building wheels
---------------

To allow for easier/more portable distribution of this tool you can build
conn-check and all its dependencies as `Python wheels <http://legacy.python.org/dev/peps/pep-0427/>`_::

    make clean-wheels
    make build-wheels
    make build-wheels-extra EXTRA=amqp
    make build-wheels-extra EXTRA=redis

The `build-wheels` make target will build conn-check and its base
dependencies, but to include the optional extra dependencies for other
checks such as amqp, redis or postgres you need to use the
`build-wheels-extra` target with the `EXTRA` env value.

By default all the wheels will be placed in `./wheels`.


Automatically generating conn-check YAML configurations
-------------------------------------------------------

The `conn-check-configs <https://pypi.python.org/pypi/conn-check-configs>`_ package contains utilities/libraries
for generating checks from existing application configurations and environments, e.g. from Django settings modules
and Juju environments.


CHANGELOG
ChangeLog for conn-check
========================

1.3.1 (2015-08-11)
------------------

 - Added guards for port numbers and the HTTP checks expected_code to cast
   any given value to an int.

1.3.0 (2015-07-15)
------------------

 - Added new conn-check-convert-fw tool to generate aws/neutron/nova/iptables rule
   commands from YAML exported by conn-check-export-fw.

1.2.0 (2015-06-19)
------------------

 - Added new smtp check to test auth/sending with SMTP servers.

1.1.0 (2015-06-05)
------------------

 - Added new conn-check-export-fw tool to export firewall egress rules in a YAML format.
 - Refactored CLI command handling code to make it easier to extend/override.

1.0.18 (2015-04-13)
-------------------

 - Ensure pyOpenSSL is always used instead of the ssl modules,
   see https://urllib3.readthedocs.org/en/latest/security.html#pyopenssl.

1.0.17 (2015-04-08)
-------------------

 - Unpin python-requests for wider distribution (e.g. precise).

1.0.16 (2015-03-06)
-------------------

 - Add --include-tags and --exclude-tags args with support for the `tags` YAML check field.

1.0.15 (2015-02-24)
-------------------

 - Package manifest fixes for debian package release.


1.0.13 (26-11-2014)
-------------------

 - Output is not buffered and ordered, with FAILED checks first, skipped last,
   and each check grouped by {type}:{host/url}.
 - TCP subchecks triggered by a HTTP check are prefixed as such.
 - There is now a -U/--unbuffered-output option to disable buffered/ordered output
   and write out to STDOUT as soon as a result is collected.

1.0.12 (17-11-2014)
-------------------

 - Command aliasing refactored, and more aliases added.

1.0.11 (04-11-2014)
-------------------

 - Disabled 30x redirects in HTTP checks by default, fixing regression introduced by requests switch.
 - Added python-requests specific options for proxy, param, cookie and auth control in HTTP checks.

1.0.10 (30-10-2014)
-------------------

 - Added a mongodb check type.

1.0.9 (23-10-2014)
------------------

 - Added --max-timeout CLI option to restrict maximum execution time.
 - Added connection timeouts to HTTP and PostgreSQL checks.
 - Added --connect-timeout CLI option to set global connection timeout.
 - Added timeout option to each individual check to override global connection timeout.

1.0.8 (22-10-2014)
------------------

 - Switched to using txrequests for HTTP requests with better proxy support.
 - Fixed UDP checks targetting host rather than IP if available.
 - Fixed initial TCP check for HTTP checks targetting upstream instead of proxy.

1.0.7 (09-10-2014)
------------------

 - Fixed HTTP proxy error in HTTP checks due to typo.

1.0.6 (06-10-2014)

 - Fixed dependencies when installing from local dir.
 - Made improvements to readme.


1.0.5 (03-10-2014)
------------------

 - Added optional headers and body arguments to HTTP checks.


1.0.4 (29-09-2014)
------------------

 - Added HTTP proxy support to http checks
 - Fixed issue with loading duplicate SSL CA certificates, and added flag to load from a custom dir


1.0.3 (24-09-2014)
------------------

 - Removed config_generators module to it's own package: conn-check-configs


1.0.2 (22-09-2014)
------------------

 - Added a script to auto-generate conn-check config YAML from a Django settings module


1.0.1 (18-09-2014)
------------------

 - Trivial release to fix setup.py tags


1.0.0 (18-09-2014)
------------------

 - Initial release
 - Broken free of UbuntuOne
 - Nagios compatible output
 - YAML configuration


conn_check/checks.py
from email.mime.text import MIMEText
import glob
import os
from pkg_resources import resource_stream
from StringIO import StringIO
import urlparse

from OpenSSL import SSL
from OpenSSL.crypto import load_certificate, FILETYPE_PEM

from twisted.internet import reactor, ssl
from twisted.internet.error import DNSLookupError, TimeoutError
from twisted.internet.abstract import isIPAddress
from twisted.internet.defer import (
    Deferred,
    inlineCallbacks,
    )
from twisted.internet.protocol import (
    ClientCreator,
    DatagramProtocol,
    Protocol,
    )
from twisted.mail.smtp import ESMTPSenderFactory
from twisted.protocols.memcache import MemCacheProtocol

from txrequests import Session
from requests.auth import HTTPDigestAuth

try:
    from requests.packages.urllib3 import disable_warnings
    from requests.packages.urllib3.contrib.pyopenssl import inject_into_urllib3
except ImportError:
    from urllib3 import disable_warnings
    from urllib3.contrib.pyopenssl import inject_into_urllib3

from .check_impl import (
    add_check_prefix,
    make_check,
    sequential_check,
    )


# Ensure we always use pyOpenSSL instead of the ssl builtin
inject_into_urllib3()

CA_CERTS = []


def load_tls_certs(path):
    cert_map = {}
    for filepath in glob.glob("{}/*.pem".format(os.path.abspath(path))):
        # There might be some dead symlinks in there,
        # so let's make sure it's real.
        if os.path.isfile(filepath):
            data = open(filepath).read()
            x509 = load_certificate(FILETYPE_PEM, data)
            # Now, de-duplicate in case the same cert has multiple names.
            cert_map[x509.digest('sha1')] = x509

    CA_CERTS.extend(cert_map.values())


class TCPCheckProtocol(Protocol):

    def connectionMade(self):
        self.transport.loseConnection()


class VerifyingContextFactory(ssl.CertificateOptions):

    def __init__(self, verify, caCerts, verifyCallback=None):
        ssl.CertificateOptions.__init__(self, verify=verify,
                                        caCerts=caCerts,
                                        method=SSL.SSLv23_METHOD)
        self.verifyCallback = verifyCallback

    def _makeContext(self):
        context = ssl.CertificateOptions._makeContext(self)
        if self.verifyCallback is not None:
            context.set_verify(
                SSL.VERIFY_PEER | SSL.VERIFY_FAIL_IF_NO_PEER_CERT,
                self.verifyCallback)
        return context


@inlineCallbacks
def do_tcp_check(host, port, tls=False, tls_verify=True,
                 timeout=None):
    """Generic connection check function."""
    if not isIPAddress(host):
        try:
            ip = yield reactor.resolve(host, timeout=(1, timeout))
        except DNSLookupError:
            raise ValueError("dns resolution failed")
    else:
        ip = host
    creator = ClientCreator(reactor, TCPCheckProtocol)
    try:
        if tls:
            context = VerifyingContextFactory(tls_verify, CA_CERTS)
            yield creator.connectSSL(ip, port, context,
                                     timeout=timeout)
        else:
            yield creator.connectTCP(ip, port, timeout=timeout)
    except TimeoutError:
        if ip == host:
            raise ValueError("timed out")
        else:
            raise ValueError("timed out connecting to {}".format(ip))


def make_tcp_check(host, port, timeout=None, **kwargs):
    """Return a check for TCP connectivity."""
    return make_check("tcp:{}:{}".format(host, port),
                      lambda: do_tcp_check(host, port, timeout=timeout),
                      info="{}:{}".format(host, port))


def make_tls_check(host, port, disable_tls_verification=False, timeout=None,
                   **kwargs):
    """Return a check for TLS setup."""

    verify = not disable_tls_verification
    check = make_check("tls:{}:{}".format(host, port),
                       lambda: do_tcp_check(host, port, tls=True,
                                            tls_verify=verify,
                                            timeout=timeout),
                       info="{}:{}".format(host, port))

    return check


class UDPCheckProtocol(DatagramProtocol):

    def __init__(self, host, port, send, expect, deferred=None,
                 timeout=None):
        self.host = host
        self.port = port
        self.send = send
        self.expect = expect
        self.deferred = deferred
        self.timeout = timeout

    def _finish(self, success, result):
        if not (self.delayed.cancelled or self.delayed.called):
            self.delayed.cancel()
        if self.deferred is not None:
            if success:
                self.deferred.callback(result)
            else:
                self.deferred.errback(result)
            self.deferred = None

    def startProtocol(self):
        self.transport.write(self.send, (self.host, self.port))
        self.delayed = reactor.callLater(self.timeout,
                                         self._finish,
                                         False, TimeoutError())

    def datagramReceived(self, datagram, addr):
        if datagram == self.expect:
            self._finish(True, True)
        else:
            self._finish(False, ValueError("unexpected reply"))


@inlineCallbacks
def do_udp_check(host, port, send, expect, timeout=None):
    """Generic connection check function."""
    if not isIPAddress(host):
        try:
            ip = yield reactor.resolve(host, timeout=(1, timeout))
        except DNSLookupError:
            raise ValueError("dns resolution failed")
    else:
        ip = host
    deferred = Deferred()
    protocol = UDPCheckProtocol(ip, port, send, expect, deferred, timeout)
    reactor.listenUDP(0, protocol)
    try:
        yield deferred
    except TimeoutError:
        if ip == host:
            raise ValueError("timed out")
        else:
            raise ValueError("timed out waiting for {}".format(ip))


def make_udp_check(host, port, send, expect, timeout=None,
                   **kwargs):
    """Return a check for UDP connectivity."""
    return make_check("udp:{}:{}".format(host, port),
                      lambda: do_udp_check(host, port, send, expect, timeout),
                      info="{}:{}".format(host, port))


def extract_host_port(url):
    parsed = urlparse.urlparse(url)
    host = parsed.hostname
    port = parsed.port
    scheme = parsed.scheme
    if not scheme:
        scheme = 'http'
    if port is None:
        if scheme == 'https':
            port = 443
        else:
            port = 80
    return host, port, scheme


def make_http_check(url, method='GET', expected_code=200, **kwargs):
    subchecks = []
    host, port, scheme = extract_host_port(url)
    proxy_url = kwargs.get('proxy_url')
    proxy_host = kwargs.get('proxy_host')
    proxy_port = int(kwargs.get('proxy_port', 8000))
    timeout = kwargs.get('timeout', None)
    expected_code = int(expected_code)

    if proxy_host:
        subchecks.append(make_tcp_check(proxy_host, proxy_port,
                                        timeout=timeout))
    else:
        subchecks.append(make_tcp_check(host, port, timeout=timeout))

    @inlineCallbacks
    def do_request():
        proxies = {}
        if proxy_url:
            proxies['http'] = proxies['https'] = proxy_url
        elif proxy_host:
            proxies['http'] = proxies['https'] = '{}:{}'.format(
                proxy_host, proxy_port)

        headers = kwargs.get('headers')
        body = kwargs.get('body')
        disable_tls_verification = kwargs.get('disable_tls_verification',
                                              False)
        allow_redirects = kwargs.get('allow_redirects', False)
        params = kwargs.get('params')
        cookies = kwargs.get('cookies')
        auth = kwargs.get('auth')
        digest_auth = kwargs.get('digest_auth')

        args = {
            'method': method,
            'url': url,
            'verify': not disable_tls_verification,
            'timeout': timeout,
            'allow_redirects': allow_redirects,
        }
        if headers:
            args['headers'] = headers
        if body:
            args['data'] = body
        if proxies:
            args['proxies'] = proxies
        if params:
            args['params'] = params
        if cookies:
            args['cookies'] = cookies
        if auth:
            args['auth'] = auth
        if digest_auth:
            args['auth'] = HTTPDigestAuth(digest_auth)

        if disable_tls_verification:
            disable_warnings()

        with Session() as session:
            request = session.request(**args)

            response = yield request
            if response.status_code != expected_code:
                raise RuntimeError(
                    "Unexpected response code: {}".format(
                        response.status_code))

    subchecks.append(make_check('', do_request,
                     info='{} {}'.format(method, url)))

    return add_check_prefix('http:{}'.format(url),
                            sequential_check(subchecks))


def make_amqp_check(host, port, username, password, use_tls=True, vhost="/",
                    timeout=None, **kwargs):
    """Return a check for AMQP connectivity."""
    from txamqp.protocol import AMQClient
    from txamqp.client import TwistedDelegate
    from txamqp.spec import load as load_spec

    subchecks = []
    subchecks.append(make_tcp_check(host, port, timeout=timeout))

    if use_tls:
        subchecks.append(make_tls_check(host, port, verify=False,
                                        timeout=timeout))

    @inlineCallbacks
    def do_auth():
        """Connect and authenticate."""
        delegate = TwistedDelegate()
        spec = load_spec(resource_stream('conn_check', 'amqp0-8.xml'))
        creator = ClientCreator(reactor, AMQClient,
                                delegate, vhost, spec)
        client = yield creator.connectTCP(host, port, timeout=timeout)
        yield client.authenticate(username, password)

    subchecks.append(make_check("amqp:{}:{}".format(host, port),
                                do_auth, info="user {}".format(username),))
    return sequential_check(subchecks)


def make_smtp_check(host, port, username, password, from_address, to_address,
                    message='', subject='', helo_fallback=False, use_tls=True,
                    timeout=None, **kwargs):
    """Return a check for SMTP connectivity."""

    subchecks = []
    subchecks.append(make_tcp_check(host, port, timeout=timeout))

    if use_tls:
        subchecks.append(make_tls_check(host, port, verify=False,
                                        timeout=timeout))

    @inlineCallbacks
    def do_connect():
        """Connect and authenticate."""
        result_deferred = Deferred()
        context_factory = None
        if use_tls:
            from twisted.internet import ssl as twisted_ssl
            context_factory = twisted_ssl.ClientContextFactory()

        body = MIMEText(message)
        body['Subject'] = subject
        factory = ESMTPSenderFactory(
            username,
            password,
            from_address,
            to_address,
            StringIO(body.as_string()),
            result_deferred,
            contextFactory=context_factory,
            requireTransportSecurity=use_tls,
            requireAuthentication=True,
            heloFallback=helo_fallback)

        if use_tls:
            reactor.connectSSL(host, port, factory, context_factory)
        else:
            reactor.connectTCP(host, port, factory)
        result = yield result_deferred

        if result[0] == 0:
            raise RuntimeError("failed to send email via smtp")

    subchecks.append(make_check("smtp:{}:{}".format(host, port),
                                do_connect, info="user {}".format(username),))
    return sequential_check(subchecks)


def make_postgres_check(host, port, username, password, database,
                        timeout=None, **kwargs):
    """Return a check for Postgres connectivity."""

    import psycopg2
    subchecks = []
    connect_kw = {
        'host': host,
        'user': username,
        'database': database,
        'connect_timeout': timeout,
    }

    if host[0] != '/':
        connect_kw['port'] = port
        subchecks.append(make_tcp_check(host, port, timeout=timeout))

    if password is not None:
        connect_kw['password'] = password

    def check_auth():
        """Try to establish a postgres connection and log in."""
        conn = psycopg2.connect(**connect_kw)
        conn.close()

    subchecks.append(make_check("postgres:{}:{}".format(host, port),
                                check_auth, info="user {}".format(username),
                                blocking=True))

    return sequential_check(subchecks)


def make_redis_check(host, port, password=None, timeout=None,
                     **kwargs):
    """Make a check for the configured redis server."""
    import txredis
    subchecks = []
    subchecks.append(make_tcp_check(host, port, timeout=timeout))

    @inlineCallbacks
    def do_connect():
        """Connect and authenticate.
        """
        client_creator = ClientCreator(reactor, txredis.client.RedisClient)
        client = yield client_creator.connectTCP(host=host, port=port,
                                                 timeout=timeout)

        if password is None:
            ping = yield client.ping()
            if not ping:
                raise RuntimeError("failed to ping redis")
        else:
            resp = yield client.auth(password)
            if resp != 'OK':
                raise RuntimeError("failed to auth to redis")

    if password is not None:
        connect_info = "connect with auth"
    else:
        connect_info = "connect"
    subchecks.append(make_check(connect_info, do_connect))

    return add_check_prefix('redis:{}:{}'.format(host, port),
                            sequential_check(subchecks))


def make_memcache_check(host, port, password=None, timeout=None,
                        **kwargs):
    """Make a check for the configured redis server."""
    subchecks = []
    subchecks.append(make_tcp_check(host, port, timeout=timeout))

    @inlineCallbacks
    def do_connect():
        """Connect and authenticate."""
        client_creator = ClientCreator(reactor, MemCacheProtocol)
        client = yield client_creator.connectTCP(host=host, port=port,
                                                 timeout=timeout)

        version = yield client.version()
        if version is None:
            raise RuntimeError('Failed retrieve memcached server version')

    subchecks.append(make_check('connect', do_connect))

    return add_check_prefix('memcache:{}:{}'.format(host, port),
                            sequential_check(subchecks))


def make_mongodb_check(host, port=27017, username=None, password=None,
                       database='test', timeout=10, **kwargs):
    """Return a check for MongoDB connectivity."""

    import txmongo
    subchecks = []
    subchecks.append(make_tcp_check(host, port, timeout=timeout))

    port = int(port)

    @inlineCallbacks
    def do_connect():
        """Try to establish a mongodb connection."""
        conn = txmongo.MongoConnection(host, port)

        conn.uri['options']['connectTimeoutMS'] = int(timeout*1000)
        if username:
            conn.uri['username'] = username
        if password:
            conn.uri['password'] = password

        # We don't start our timeout callback until now, otherwise we might
        # elapse part of our timeout period during the earlier TCP check
        reactor.callLater(timeout, timeout_handler)

        mongo = yield conn
        names = yield mongo[database].collection_names()
        if names is None:
            raise RuntimeError('Failed to retrieve collection names')

    def timeout_handler():
        """Manual timeout handler as txmongo timeout args don't always work."""
        if 'deferred' in do_connect.func_dict:
            err = ValueError("timeout connecting to mongodb")
            do_connect.func_dict['deferred'].errback(err)

    if any((username, password)):
        connect_info = "connect with auth"
    else:
        connect_info = "connect"
    subchecks.append(make_check(connect_info, do_connect))

    return add_check_prefix('mongodb:{}:{}'.format(host, port),
                            sequential_check(subchecks))


CHECKS = {
    'tcp': {
        'fn': make_tcp_check,
        'args': ['host', 'port'],
    },
    'tls': {
        'fn': make_tls_check,
        'args': ['host', 'port'],
    },
    'udp': {
        'fn': make_udp_check,
        'args': ['host', 'port', 'send', 'expect'],
    },
    'http': {
        'fn': make_http_check,
        'args': ['url'],
    },
    'amqp': {
        'fn': make_amqp_check,
        'args': ['host', 'port', 'username', 'password'],
    },
    'postgres': {
        'fn': make_postgres_check,
        'args': ['host', 'port', 'username', 'password', 'database'],
    },
    'redis': {
        'fn': make_redis_check,
        'args': ['host', 'port'],
    },
    'memcache': {
        'fn': make_memcache_check,
        'args': ['host', 'port'],
    },
    'mongodb': {
        'fn': make_mongodb_check,
        'args': ['host'],
    },
    'smtp': {
        'fn': make_smtp_check,
        'args': ['host', 'port', 'username', 'password'],
    },
}

CHECK_ALIASES = {
    'mongo': 'mongodb',
    'memcached': 'memcache',
    'ssl': 'tls',
    'postgresql': 'postgres',
    'https': 'http',
}
